### PR TITLE
[DD4hep] always close geometry

### PR DESCRIPTION
**Pull request by ianna**
#### PR description:

* Fixes `MagneticField/Engine` unit tests
* Requires https://github.com/cms-sw/cmsdist/pull/6353

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by ianna** [comment]
please test with https://github.com/cms-sw/cmsdist/pull/6353

**Comment by ianna** [comment]
@civanch - FYI

**Comment by ianna** [comment]
@mrodozov and @smuzaffar - FYI

**Comment by civanch** [comment]
@ianna , one of crashes:

%MSG-e TkDetLayers:   TSGFromL2Muon:hltL3NoFiltersNoVtxTrajSeedOIHit@streamBeginRun  02-Nov-2020 22:00:27 CET Run: 1 Stream: 0
 ForwardDiskSectorBuilderFromDet: Trying to build Petal Wedge from Dets at different z positions !! Delta_z = -0.951241
%MSG
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 02-Nov-2020 22:00:29.869 CET

**Comment by ianna** [comment]
> @ianna , one of crashes:
> 
> %MSG-e TkDetLayers: TSGFromL2Muon:hltL3NoFiltersNoVtxTrajSeedOIHit@streamBeginRun 02-Nov-2020 22:00:27 CET Run: 1 Stream: 0
> ForwardDiskSectorBuilderFromDet: Trying to build Petal Wedge from Dets at different z positions !! Delta_z = -0.951241
> %MSG
> Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 02-Nov-2020 22:00:29.869 CET

I don’t think it’s related to the PR. This should not have any effect on the regular workflows.

**Comment by smuzaffar** [comment]
@ianna , root update fixed the unit test issue but now we have a new unit test `testDD4hepDDSolid` failing [a]. Do you think it is something in cmssw to fix it? If yes then may be we can go ahead with root update and include it for next IB 

[a]
```
world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs1_1
#78: trunctubs1[1] is a TruncatedTube == TruncatedTube
world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs2_1
#79: trunctubs2[1] is a TruncatedTube == TruncatedTube
world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:V_1101_1333
#80: V_1101[1333] is a  == Solid not initialized
F

DDSolid.cppunit.cc:46:Assertion
Test name: testDDSolid::checkDDSolid
assertion failed
- Expression: title.compare(name) == 0

Failures !!!
Run: 1   Failure total: 1   Failures: 1   Errors: 0

---> test testDD4hepDDSolid had ERRORS
```

**Comment by ianna** [comment]
> @ianna , root update fixed the unit test issue but now we have a new unit test `testDD4hepDDSolid` failing [a]. Do you think it is something in cmssw to fix it? If yes then may be we can go ahead with root update and include it for next IB
> 
> [a]
> 
> ```
> world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs1_1
> #78: trunctubs1[1] is a TruncatedTube == TruncatedTube
> world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs2_1
> #79: trunctubs2[1] is a TruncatedTube == TruncatedTube
> world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:V_1101_1333
> #80: V_1101[1333] is a  == Solid not initialized
> F
> 
> DDSolid.cppunit.cc:46:Assertion
> Test name: testDDSolid::checkDDSolid
> assertion failed
> - Expression: title.compare(name) == 0
> 
> Failures !!!
> Run: 1   Failure total: 1   Failures: 1   Errors: 0
> 
> ---> test testDD4hepDDSolid had ERRORS
> ```

@smuzaffar - are you sure the patch for title is in the updated root? It supposed to fix this unit test - the titles were not copied for reflection solids.

**Comment by ianna** [comment]
> > @ianna , root update fixed the unit test issue but now we have a new unit test `testDD4hepDDSolid` failing [a]. Do you think it is something in cmssw to fix it? If yes then may be we can go ahead with root update and include it for next IB
> > [a]
> > ```
> > world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs1_1
> > #78: trunctubs1[1] is a TruncatedTube == TruncatedTube
> > world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs2_1
> > #79: trunctubs2[1] is a TruncatedTube == TruncatedTube
> > world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:V_1101_1333
> > #80: V_1101[1333] is a  == Solid not initialized
> > F
> > 
> > DDSolid.cppunit.cc:46:Assertion
> > Test name: testDDSolid::checkDDSolid
> > assertion failed
> > - Expression: title.compare(name) == 0
> > 
> > Failures !!!
> > Run: 1   Failure total: 1   Failures: 1   Errors: 0
> > 
> > ---> test testDD4hepDDSolid had ERRORS
> > ```
> 
> @smuzaffar - are you sure the patch for title is in the updated root? It supposed to fix this unit test - the titles were not copied for reflection solids.

https://github.com/root-project/root/pull/6736 should fix `testDD4hepDDSolid` unit test

**Comment by ianna** [comment]
> > > @ianna , root update fixed the unit test issue but now we have a new unit test `testDD4hepDDSolid` failing [a]. Do you think it is something in cmssw to fix it? If yes then may be we can go ahead with root update and include it for next IB
> > > [a]
> > > ```
> > > world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs1_1
> > > #78: trunctubs1[1] is a TruncatedTube == TruncatedTube
> > > world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:trunctubs2_1
> > > #79: trunctubs2[1] is a TruncatedTube == TruncatedTube
> > > world_volume/testLogicalParts:MotherOfAllBoxes_2/testLogicalParts:V_1101_1333
> > > #80: V_1101[1333] is a  == Solid not initialized
> > > F
> > > 
> > > DDSolid.cppunit.cc:46:Assertion
> > > Test name: testDDSolid::checkDDSolid
> > > assertion failed
> > > - Expression: title.compare(name) == 0
> > > 
> > > Failures !!!
> > > Run: 1   Failure total: 1   Failures: 1   Errors: 0
> > > 
> > > ---> test testDD4hepDDSolid had ERRORS
> > > ```
> > 
> > 
> > @smuzaffar - are you sure the patch for title is in the updated root? It supposed to fix this unit test - the titles were not copied for reflection solids.
> 
> [root-project/root#6736](https://github.com/root-project/root/pull/6736) should fix `testDD4hepDDSolid` unit test

@mrodozov - https://github.com/root-project/root/pull/6738 was not merged - does it mean the patch is not in 6.22 version?

**Comment by mrodozov** [comment]
@ianna  I took the PR in question as cms-sw/root#145 and tested with it so it is taken on top of our existing ROOT 6-22.
Also 32000 is taken 
Sorry for this delay it's just logs load a bit slow , but it basically shows this:
```
20:21:46 + PULL_REQUESTS='cms-sw/cmsdist#6364 cms-sw/root#145 cms-sw/cmssw#32000'
20:21:46 ++ echo cms-sw/cmsdist#6364 cms-sw/root#145 cms-sw/cmssw#32000
```
which is the extra PRs taken. 145 on ROOT is the patch in question. I'll post the patching part from the jenkins log the moment it loads.

**Comment by smuzaffar** [comment]
@ianna , I confirm that our tests indeed included that change [a]. Please note that with this change it fixed the unit tests (MagneticFieldEngineTestDriver ) which is failing in IB but now a new unit tests testDD4hepDDSolid failed.


[a]
```
From git://github.com/mrodozov/root
 * branch            pick-fix-over-622-in-cmsdistmaster -> FETCH_HEAD
Updating c8fedfb..7eb659e
Fast-forward
 geom/geom/src/TGeoNode.cxx   | 2 ++
 geom/geom/src/TGeoVolume.cxx | 2 ++
 2 files changed, 4 insertions(+)

```

**Comment by ianna** [comment]
> @ianna , I confirm that our tests indeed included that change [a]. Please note that with this change it fixed the unit tests (MagneticFieldEngineTestDriver ) which is failing in IB but now a new unit tests testDD4hepDDSolid failed.
> 
> [a]
> 
> ```
> From git://github.com/mrodozov/root
>  * branch            pick-fix-over-622-in-cmsdistmaster -> FETCH_HEAD
> Updating c8fedfb..7eb659e
> Fast-forward
>  geom/geom/src/TGeoNode.cxx   | 2 ++
>  geom/geom/src/TGeoVolume.cxx | 2 ++
>  2 files changed, 4 insertions(+)
> ```

@smuzaffar - correction - this PR fixes the unit tests (MagneticFieldEngineTestDriver ) and root fix suppose to fix the unit test `testDD4hepDDSolid`...

**Comment by smuzaffar** [comment]
ah OK understood.
Looks like root has merged  root-project/root#6738 now which I am testing via https://github.com/cms-sw/cmsdist/pull/6367 . Please do not use https://github.com/cms-sw/cmsdist/pull/6367 to test your PR. let the cmsdist test finish.

**Comment by smuzaffar** [comment]
@ianna , the change you need for root is now available in ROOT622 IBs (https://github.com/cms-sw/root/commits/cms/v6-22-00-patches/50e52ba43c ). Can you please check latest ROOT622 IB and see if root fix was enough to fix the unit test?

**Comment by ianna** [comment]
> @ianna , the change you need for root is now available in ROOT622 IBs (https://github.com/cms-sw/root/commits/cms/v6-22-00-patches/50e52ba43c ). Can you please check latest ROOT622 IB and see if root fix was enough to fix the unit test?

Thanks, I'll do it asap

**Comment by ianna** [comment]
@smuzaffar - tested against `CMSSW_11_2_ROOT622_X_2020-11-04-2300` - unfortunately, it does not fix it :-(
This unit tests points to a problem in ROOT. Do we want to silence it?

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
@ianna 
shall we merge this PR together with the fix of the unit test testDD4hepDDSolid?

**Comment by ianna** [comment]
@civanch and @smuzaffar - it needs one more patch from ROOT: https://github.com/root-project/root/pull/6778

**Comment by smuzaffar** [comment]
@mrodozov , can you please get this root PR for our v6.22 branch so that we can test it

**Comment by mrodozov** [comment]
sure, I'm getting it

**Comment by smuzaffar** [comment]
@ianna , it still fails. This is now tested using https://github.com/cms-sw/root/commits/cms/v6-22-00-patches/d6156de where we have both 
- https://github.com/cms-sw/root/commit/76b0e9d16e71f7636b0503c68c289982c5205a58
- https://github.com/cms-sw/root/commit/76b0e9d16e71f7636b0503c68c289982c5205a58

**Comment by smuzaffar** [comment]
please test with  cms-sw/cmsdist#6382,cms-sw/cmsdist#6391

**Comment by silviodonato** [comment]
@ianna 
both https://github.com/cms-sw/cmsdist/pull/6382
and https://github.com/cms-sw/cmsdist/pull/6391
have been merged.
Currently, the `MagneticField/Engine` unit test in `11_2_X` (https://github.com/cms-sw/cmssw/issues/31987).
This PR fixes the unit test failure in `MagneticField/Engine` but it gets a new failure in the`DetectorDescription/DDCMS` unit test related to the "title" comparison and it can be ignored for the time being, if I understood correctly from the sim meeting.
```
DDSolid.cppunit.cc:46:Assertion
Test name: testDDSolid::checkDDSolid
assertion failed
- Expression: title.compare(name) == 0

Failures !!!
Run: 1   Failure total: 1   Failures: 1   Errors: 0

---> test testDD4hepDDSolid had ERRORS
```

So it looks like it is convenient to merge this PR in 11_2_0_pre9

**Comment by ianna** [comment]
> @ianna
> both [cms-sw/cmsdist#6382](https://github.com/cms-sw/cmsdist/pull/6382)
> and [cms-sw/cmsdist#6391](https://github.com/cms-sw/cmsdist/pull/6391)
> have been merged.
> Currently, the `MagneticField/Engine` unit test in `11_2_X` (#31987).
> This PR fixes the unit test failure in `MagneticField/Engine` but it gets a new failure in the`DetectorDescription/DDCMS` unit test related to the "title" comparison and it can be ignored for the time being, if I understood correctly from the sim meeting.
> 
> ```
> DDSolid.cppunit.cc:46:Assertion
> Test name: testDDSolid::checkDDSolid
> assertion failed
> - Expression: title.compare(name) == 0
> 
> Failures !!!
> Run: 1   Failure total: 1   Failures: 1   Errors: 0
> 
> ---> test testDD4hepDDSolid had ERRORS
> ```
> 
> So it looks like it is convenient to merge this PR in 11_2_0_pre9

Yes, the unit test points to a title issue. I think, we can ignore it for the moment. I'll pin the ROOT team again. Thanks!

**Comment by ianna** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Comment by silviodonato** [comment]
merge

**Comment by bsunanda** [comment]
@ianna Are we now using the new dd4hep code for reflected solids and not ROOT ones?

**Comment by ianna** [comment]
> @ianna Are we now using the new dd4hep code for reflected solids and not ROOT ones?

No, not yet. This TGeo reflection factory call should be disabled/removed:
https://github.com/cms-sw/cmssw/blob/89173531c6e678141df030bf97413beaaa0757ff/DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc#L2249
 @ghugo83 is looking into this.

**Comment by ghugo83** [comment]
Yes if the test of the reflections in CSCs in the latest DD4hep @MarkusFrankATcernch turns out to not show any surprising regression, I will deactivate the TGeo reflection.

**Comment by bsunanda** [comment]
@ianna @ghugo83 It will be good to test after the new dd4hep is integrated and use the PR which disables this. Could you please let me know once the PR is available

**Comment by ghugo83** [comment]
@bsunanda The PR is straightforward, ok let me do it now then.
By the way, I noticed there are still overlaps in ECAL preshower with DD4hep ? - off-topic from this point though.

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc [hunk 1]
```diff
@@ -2058,7 +2058,7 @@ static long load_dddefinition(Detector& det, xml_h element) {
 
     string fname = xml::DocumentHandler::system_path(element);
     bool open_geometry = dddef.hasChild(DD_CMU(open_geometry)) ? dddef.child(DD_CMU(open_geometry)) : true;
-    bool close_geometry = dddef.hasChild(DD_CMU(close_geometry)) ? dddef.hasChild(DD_CMU(close_geometry)) : false;
+    bool close_geometry = dddef.hasChild(DD_CMU(close_geometry)) ? dddef.hasChild(DD_CMU(close_geometry)) : true;
```
Comment (ghugo83) [inline DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:2061]: @ianna why not 
`bool close_geometry = true` ?
This must be equivalent.
Or did you mean:
 `bool close_geometry = dddef.hasChild(DD_CMU(close_geometry)) ? dddef.**child**(DD_CMU(close_geometry)) : true;`
instead?

**Comment by ghugo83** [comment]
@bsunanda Handled here: https://github.com/cms-sw/cmssw/pull/32112

File: DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc (hunk 1 above)
Comment (ianna) [reply DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc:2061]: Closing (or not) geometry should be controlled via configuration, so it should be: 
```c++
bool close_geometry = dddef.hasChild(DD_CMU(close_geometry)) ? dddef.hasChild(DD_CMU(close_geometry)) : false;
```
but there are places in the configurations where the `close_geometry` tag is not present, but needed. For example, in the MagneticField geometry. It needs a special "hidden" volume.

**Comment by ianna** [comment]
> @ianna
> both [cms-sw/cmsdist#6382](https://github.com/cms-sw/cmsdist/pull/6382)
> and [cms-sw/cmsdist#6391](https://github.com/cms-sw/cmsdist/pull/6391)
> have been merged.
> Currently, the `MagneticField/Engine` unit test in `11_2_X` (#31987).
> This PR fixes the unit test failure in `MagneticField/Engine` but it gets a new failure in the`DetectorDescription/DDCMS` unit test related to the "title" comparison and it can be ignored for the time being, if I understood correctly from the sim meeting.
> 
> ```
> DDSolid.cppunit.cc:46:Assertion
> Test name: testDDSolid::checkDDSolid
> assertion failed
> - Expression: title.compare(name) == 0
> 
> Failures !!!
> Run: 1   Failure total: 1   Failures: 1   Errors: 0
> 
> ---> test testDD4hepDDSolid had ERRORS
> ```
> 
> So it looks like it is convenient to merge this PR in 11_2_0_pre9

fixed via https://github.com/cms-sw/cmssw/pull/32112

**Pull request merged**